proxy/coco/src/peer/run_state.rs

```diff
@@ -278,45 +381,87 @@ impl RunState {
 
                 vec![]
             },
-            // Announce new updates while the peer is online.
-            (
-                Status::Online(_) | Status::Started(_) | Status::Syncing(_, _),
-                Event::Announce(AnnounceEvent::Tick),
-            ) => vec![Command::Announce],
-            // Query requested URNs while online.
-            (
-                Status::Online(_) | Status::Syncing(_, _),
-                Event::Request(RequestEvent::Query(urn)),
-            ) => vec![RequestCommand::Query(urn.clone()).into()],
-            // Clone requested URLs while online.
-            (
-                Status::Online(_) | Status::Syncing(_, _),
-                Event::Request(RequestEvent::Clone(url)),
-            ) => vec![RequestCommand::Clone(url.clone()).into()],
             // Found URN.
             (
                 _,
-                Event::Protocol(ProtocolEvent::Gossip(Info::Has(Has {
+                ProtocolEvent::Gossip(Info::Has(Has {
                     provider,
                     val: Gossip { urn, .. },
-                }))),
-            ) => vec![RequestCommand::Found(RadUrl {
-                authority: provider.peer_id.clone(),
-                urn: urn.clone(),
-            })
-            .into()],
+                })),
+            ) => {
+                // FIXME(xla): That the result is disregarded is dangerous, as either the surface of
+                // the WaitingRoom should be adjusted or the Result handled explicitely.
+                self.waiting_room
+                    .found(
+                        RadUrl {
+                            urn,
+                            authority: provider.peer_id,
+                        },
+                        Instant::now(),
+                    )
+                    .ok();
+                vec![]
+            },
             _ => vec![],
-        };
+        }
+    }
 
-        log::trace!(
-            "TRANSITION: {:?} -> [{:?} -> {:?}] -> {:?}",
-            event,
-            old_status,
-            self.status,
-            cmds
-        );
+    /// Handle [`RequestInput`]s.
+    fn handle_request(&mut self, input: RequestInput) -> Vec<Command> {
+        match (&self.status, input) {
+            // Check for new querie and clone requests.
+            (Status::Online(_) | Status::Syncing(_, _), RequestInput::Tick) => {
+                let mut cmds = Vec::with_capacity(2);
 
-        cmds
+                if let Some(urn) = self.waiting_room.next_query(Instant::now()) {
+                    cmds.push(Command::Request(RequestCommand::Query(urn)));
+                }
+                if let Some(url) = self.waiting_room.next_clone() {
+                    cmds.push(Command::Request(RequestCommand::Clone(url)));
+                }
+                cmds
+            },
+            // FIXME(xla): Come up with a strategy for the results returned by the waiting room.
```

Comment:
@FintanH What's ur take on the Results here and how to preserve semantics?

Reply:
A very important `Error` result here is the `TimedOut` variant. We will likely want to communicate that back. I'm almost thinking that we could just communicate the `Result` back on the channel so that the endpoint can also communicate that error.

What do you think?

Reply:
Could you expand which of the calls below returns `TimedOut`? Sounds almost more important to react to the returns inline here and change the state machine semantics. Currently the only input that carries a channel is the `Requested` variant.

Reply:
> Could you expand which of the calls below returns TimedOut?

`queired` and `cloning` do -- check out their docs

> Sounds almost more important to react to the returns inline here and change the state machine semantics. Currently the only input that carries a channel is the Requested variant.

Ah right, wasn't thinking through fully. It's more like that we'll want to report the `TimedOut` errors when we hook up notifications -- as well as the successful cases of "found peer" and "cloned repo".

On top of that, logging warnings could be useful to know when we're requesting things with the wrong state transitions.